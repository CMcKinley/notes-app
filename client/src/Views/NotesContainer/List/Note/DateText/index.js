import React, { memo } from "react";
import PropTypes from "prop-types";
import moment from "moment";

const DateText = ({ dateToDisplay, text }) => {
  const formattedDate = dateToDisplay
    ? moment(dateToDisplay).format("MMMM, DD, YYYY")
    : null;
  const formattedTime = dateToDisplay
    ? moment(dateToDisplay).format("h:mma")
    : null;
  const today = moment().format("MMMM, DD, YYYY");

  return (
    <span>
      {text} {today === formattedDate ? "Today" : formattedDate} at{" "}
      {formattedTime}
    </span>
  );
};

DateText.propTypes = {
  text: PropTypes.string,
  dateToDisplay: PropTypes.string
};

DateText.defaultProps = {
  text: "",
  dateToDisplay: ""
};

export default memo(DateText);
