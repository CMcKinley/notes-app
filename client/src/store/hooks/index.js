export { default as useGetNotes } from './useGetNotes';
export { default as useEditNotes } from './useEditNotes';
export { default as useAddNotes } from './useAddNotes';