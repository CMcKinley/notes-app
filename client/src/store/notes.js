import { handleActions, createAction } from "redux-actions";
import cloneDeep from "lodash.clonedeep";

const defultState = {
  noteIds: [],
  noteMap: {},
  noteToEdit: null
};

const prefix = "NOTES/";

// Action Creators
export const setNotes = createAction(`${prefix}SET_NOTES`);
export const editNoteInStore = createAction(`${prefix}EDIT_NOTE`);
export const addNoteToStore = createAction(`${prefix}ADD_NOTE`);
export const setNoteToEdit = createAction(`${prefix}SET_NOTE_TO_EDIT`);

// Reducer
export default handleActions(
  {
    // Load notes into state. Notes are stored as an array of ids and an object where each key is the note's id and the value is the note object itself.
    [setNotes]: (state, { payload }) => {
      const notes = payload || [];
      const noteIds = [];
      const noteMap = {};

      notes.forEach(note => {
        const { id } = note || {};
        noteIds.push(id);
        noteMap[id] = note;
      });

      return { ...state, noteIds, noteMap };
    },

    // Edit an existing note
    [editNoteInStore]: (state, { payload }) => {
      const { id } = payload || {};
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const newMap = cloneDeep(state.noteMap);
      newMap[id] = { ...newMap[id], ...payload };
      return { ...state, noteMap: newMap };
    },

    // Add a new note to the id array and note map
    [addNoteToStore]: (state, { payload }) => {
      const { id } = payload || {};
      const noteIds = [...state.noteIds];
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const noteMap = cloneDeep(state.noteMap);
      // Add note to beginning of array
      noteIds.unshift(payload.id);
      noteMap[id] = payload;
      return { ...state, noteIds, noteMap };
    },

    // Set the note that is to be edited in store
    [setNoteToEdit]: (state, { payload }) => ({ ...state, noteToEdit: payload })
  },
  defultState
);
