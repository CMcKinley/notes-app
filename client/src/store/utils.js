import axios from "axios";

// Store tools will be used to make http calls to the node api
export default class StoreTools {
  async base(url, body, method = "get", headers = {}) {
    let config = { url: `/api/v1/${url}`, method, headers };
    if (body) {
      config.data = body;
    }
    return axios(config);
  }
  // GET requests
  async get(url, headers = {}) {
    return this.base(url, null, "get", headers);
  }
  // PUT requests
  put(url, body, headers = {}) {
    return this.base(url, body, "put", headers);
  }
  // POST requests
  post(url, body, headers = {}) {
    return this.base(url, body, "post", headers);
  }
}
