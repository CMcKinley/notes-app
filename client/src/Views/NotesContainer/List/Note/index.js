import React, { memo } from "react";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";
import { makeStyles } from "@material-ui/core/styles";
import moment from "moment";
import IconButton from "@material-ui/core/IconButton";
import EditIcon from "@material-ui/icons/Create";
import { setNoteToEdit } from "store/notes.js";
import { useDispatch } from "react-redux";

const useStyles = makeStyles(() => ({
  dateText: {
    fontSize: ".7em"
  },
  bottom: {
    width: "100%",
    display: "flex",
    justifyContent: "space-between"
  },
  note: {
    width: "100%",
    margin: "8px 0 16px 0"
  },
  content: {
    display: "flex",
    width: "100%",
    justifyContent: "space-between"
  },
  listItem: {
    padding: "8px 16px",
    display: "flex",
    flexDirection: "column",
    width: "100%",
    alignItems: "center",
    transition: "all .3s ease",
    borderRadius: "none",
    "&:hover": { background: "rgba(0,0,0,0.1)" }
  }
}));

const Note = ({ noteId }) => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const { text, created, updated, id } = useSelector(
    state => state.notes.noteMap[noteId]
  );

  const formattedDate = moment(created).format("MMMM, DD, YYYY");
  const formattedUpdateDate = updated
    ? moment(updated).format("MMMM, DD, YYYY")
    : null;
  const today = moment(created).format("MMMM, DD, YYYY");

  return (
    <li className={classes.listItem}>
      <div className={classes.content}>
        <div className={classes.note}>{text}</div>
        <IconButton onClick={() => dispatch(setNoteToEdit(id))}>
          <EditIcon fontSize="small" />
        </IconButton>
      </div>
      <div className={`${classes.dateText} ${classes.bottom}`}>
        <span>
          {today === formattedDate ? "Today" : formattedDate} at{" "}
          {moment(created).format("h:mma")}
        </span>
        {updated && (
          <span>
            Edited{" "}
            {today === formattedUpdateDate ? "today" : formattedUpdateDate} at{" "}
            {moment(created).format("h:mma")}
          </span>
        )}
      </div>
    </li>
  );
};

Note.propTypes = {
  noteId: PropTypes.string
};

Note.defaultProps = {
  noteId: ""
};

export default memo(Note);
