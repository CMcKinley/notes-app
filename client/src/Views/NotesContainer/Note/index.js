import React, { memo, useMemo } from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useSelector, shallowEqual } from "react-redux";
import { makeStyles } from "@material-ui/core/styles";
import IconButton from "@material-ui/core/IconButton";
import EditIcon from "@material-ui/icons/Create";
import { setNoteToEdit } from "store/notes.js";
import { useDispatch } from "react-redux";
import DateText from "./DateText";

const useStyles = makeStyles(() => ({
  dateText: {
    fontSize: ".7em"
  },
  bottom: {
    width: "100%",
    display: "flex",
    justifyContent: "space-between"
  },
  note: {
    width: "100%",
    margin: "8px 0 16px 0",
    cursor: "pointer"
  },
  content: {
    display: "flex",
    width: "100%",
    justifyContent: "space-between"
  },
  listItem: {
    padding: "8px 16px",
    display: "flex",
    flexDirection: "column",
    width: "100%",
    alignItems: "center",
    transition: "all .3s ease",
    borderRadius: "none",
    "&:hover": { background: "rgba(0,0,0,0.1)" }
  },
  editIcon: {
    height: "44px"
  }
}));

const Note = ({ noteId, history }) => {
  const classes = useStyles();
  const dispatch = useDispatch();

  // Get note details from store by the noteId prop
  const { text, created, updated, id } = useSelector(
    state => state.notes.noteMap[noteId] || {},
    shallowEqual
  );

  // Render note list item
  return (
    <li className={classes.listItem}>
      <div className={classes.content}>
        <div className={classes.note} onClick={() => history.push(`/${id}`)}>
          {text}
        </div>
        <IconButton
          className={classes.editIcon}
          onClick={() => dispatch(setNoteToEdit(id))}
        >
          <EditIcon fontSize="small" />
        </IconButton>
      </div>
      <div className={`${classes.dateText} ${classes.bottom}`}>
        <DateText dateToDisplay={created} />
        {updated && <DateText text="Edited" dateToDisplay={updated} />}
      </div>
    </li>
  );
};

Note.propTypes = {
  noteId: PropTypes.string,
  history: PropTypes.object
};

Note.defaultProps = {
  noteId: "",
  history: {}
};

export default memo(withRouter(Note));
