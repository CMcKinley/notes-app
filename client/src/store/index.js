import { combineReducers, createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import notes from "./notes.js";

const store = combineReducers({ notes });

const composeEnhancers = composeWithDevTools({});

export default createStore(store, composeEnhancers(applyMiddleware(thunk)));
