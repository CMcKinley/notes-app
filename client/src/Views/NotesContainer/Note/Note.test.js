import React from "react";
import renderer from "react-test-renderer";
import Note from "./index.js";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { defultState } from "store/notes";
import { BrowserRouter as Router } from "react-router-dom";
export const mockStore = configureStore([]);

const initialState = { notes: defultState };
const store = mockStore(initialState);

describe("Note", () => {
  it("Renders correctly", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Note />
        </Router>
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
