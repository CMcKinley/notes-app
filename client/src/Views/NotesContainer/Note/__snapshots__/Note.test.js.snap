// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Renders correctly 1`] = `
<li
  className="makeStyles-listItem-5"
>
  <div
    className="makeStyles-content-4"
  >
    <div
      className="makeStyles-note-3"
      onClick={[Function]}
    />
    <button
      className="MuiButtonBase-root MuiIconButton-root makeStyles-editIcon-6"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="makeStyles-dateText-1 makeStyles-bottom-2"
  >
    <span>
      
       
       at
       
    </span>
  </div>
</li>
`;
